Improve parse performance by using a faster hash function

rcombs opened the pull request:
This dramatically reduces the number of muls we end up doing, which improves performance on cache-hash-heavy files quite a bit; I see nearly 40% on a very drawing-heavy test. I expect perf to improve a bit more once @MrSmile's strdup PR lands and the `strlen` is no longer used (in addition to the perf benefits that change provides alone).

MrSmile reviewed (commented):
I'm always welcoming performance, but this hash calculation is quite complicated. I think it can be further simplified without impact on quality or performance. For example, instead of ton of ifs at the start of `wyhash()` function isn't something like
```c
uint64_t a = 0;
memcpy(&a, key, len);
a = bswap_if_needed(a);
```
is better? Also `bswap` is needed only if we care about repeatability of hash values under big-endian platforms (are we even target big-endian?)

TheOneric commented:
> (are we even target big-endian?)

libass is currently being built for some big-endian architectures by at 
least Debian: https://buildd.debian.org/status/package.php?p=libass&suite=sid
So it appears big-endian libass is currently working.

astiob commented on libass/ass_utils.h:
```diff
@@ -30,6 +30,7 @@
 #include <math.h>
 
 #include "ass.h"
+#include "wyhash.h"
```
The file is a verbatim copy of the upstream/official wyhash code. Of course, its licence permits us to change it however we see fit, but personally, I prefer to keep such code intact (unless they have bugs that need fixing), e. g. so that it can be easily updated when upstream code changes.

MrSmile replied:
Yes, I'm usually in favor of keeping external dependencies intact too. But this specific file looks like code written by inexperienced student and it's quite hard to understand what's going on in that. So I prefer to borrow algorithm only and not the source code.

I'm planning to check original upstream repo and try to simplify code and maybe even improve performance.

TheOneric replied:
As upstream appears to be active and even more so with MrSmile planning to submit improvements to upstream, using the header verbatim for now seems imho like the better option.

MrSmile replied:
Sadly, upstream has research code instead of production (all that `WYHASH_CONDOM`, `WYHASH_32BIT_MUM`, bunch of auxiliary functions) so that is unlikely to happen. I've done some experiments on wyhash [here](https://github.com/MrSmile/wyhash/tree/test) and it's indeed one of the fastest on bulk data. Reducing the number of multiply operations per byte can improve bulk performance, but I've not found a way to do so without sacrificing hash quality. Performance on small fixed-size elements can be improved but it's not the problem case.

astiob reviewed (commented):
I trust you have tested this and found this improve overall performance significantly. (While looking up wyhash, I saw some research claiming hash functions that are conventionally “slower” but have shorter code perform better in hash tables in practice. But perhaps the scenarios where it matters to libass involve long vector drawing texts, whereas the research probably focussed on shorter keys such as field names in dynamic languages.)

Remind me: have you tried other hash functions or just wyhash?

Looking up wyhash also reminded me that we can probably extract another, possibly even more significant, performance improvement by changing the _structure_ of our hash tables.

rcombs commented on libass/wyhash.h:
```diff
@@ -0,0 +1,115 @@
+#ifndef WYHASH_CONDOM
+#define WYHASH_CONDOM 1 //0: read 8 bytes before and after boundaries, dangerous but fastest. 1: normal valid behavior 2: extra protection against entropy loss (probability=2^-63), aka. "blind multiplication"
+#endif
+#define WYHASH_32BIT_MUM 0	//faster on 32 bit system
```
Dunno; I'm not especially concerned about perf on 32-bit anymore, so I wasn't too worried either way.

rcombs commented on libass/ass_utils.h:
```diff
@@ -160,32 +161,15 @@ static inline int double_to_d22(double x)
-    }
-
-    return hval;
+    static const uint64_t _wyp[5] = {0xa0761d6478bd642full, 0xe7037ed1a0b428dbull, 0x8ebc6af09c88c6e3ull, 0x589965cc75374cc3ull, 0x1d8e4e27c47d124full};
```
I got this from upstream wyhash; not sure if it's essential to it passing smhasher so I figured I'd just keep it as-is.

TheOneric replied:
Current wyhash-upstream defines this default key in the header, making this definition unneccessary. *(There's also a function to generate other keys, haven't tested other keys though)*

https://github.com/wangyi-fudan/wyhash/blob/41043dc14fd4995e5e343c21cc14b5b075a9de63/wyhash.h#L116

The version in this pr, doesn't have this yet. Any particular reason you didn't use the most recent wyhash.h?

astiob replied:
The parameter is named `secret` if I’m not mistaken and the paper (in the upstream repo) calls it a “secret”, and it’s a parameter rather than just an internal constant… but then the current header has a mysterious `wyhash64` function that uses that value as a constant and seemingly so does `wyrand`… I don’t even mind the dense code so much, but I sure wish this was better _documented._

TheOneric replied:
I don't see a reason why this parameter should need to be secret; wyhash is not a cryptographic hash so even if this value were truly secret it would be unsuitable for security relevant use-case.
And as for generating a unique "secret": albeit referring to an older version, [wyhash#53](https://github.com/wangyi-fudan/wyhash/issues/53) suggests, that even when using `make_secret()` it takes many tries to obtain a usable "secret". Imho using the default `_wyp`-"secret" seems like the best option.

MrSmile replied:
It's indeed better to use unique secret for each run (per ASS_Library or Cache?). But upstream secret generation code is cumbersome and slow, not useful in runtime context. Why even secret values should be prime numbers? Even if they should be primes, there are much better algorithms for that. Also there is not much sense in making secret generation function inline and polluting the header, it's not in a fast path anyway.

rcombs reviewed (commented):
I did try CRC32 (using the x86 instruction for it) and it wasn't nearly as fast as wyhash, somewhat surprisingly. Might've helped if I'd used larger word sizes, but regardless it would've required dispatch between plain-C and x86-with-CRC variants.

One other idea to help with these files: we should probably only hash the first, say, 1024 bytes of any given string.

TheOneric reviewed (commented):
Brings a nice perf improvements in some cases, ~~though I didn't get near the 45% mentioned on IRC – was this a different file or was the contribution of the now merged strdup-reduction just that much larger? *(Or perhaps the systematic difference between callgrind and timed measurement is much larger than anticipated ...)*~~

```
              Sample    Overall Gain [%]
            dialogue	 2.86
            fxkara_1	 0.87
            mpv-7712	 2.63
  zoinks-neptunia-op	 6.64
   YuruCamp2_02_rock	23.14
      1eventPerFrame	 0.01
    32eventsPerFrame	 0.23
    64eventsPerFrame	 0.36
   128eventsPerFrame	 0.42
   512eventsPerFrame	 0.55
```

Total performance gains were measured using Callgrind.
`event(s)PerFrame` samples are synthetic; other samples are taken from real-world files and measured with real-world fps.

rcombs commented:
That was on this file (double-compressed for github):
[YuruCamp2_02_-_TS_-_rock.ass.xz.zip](https://github.com/libass/libass/files/6030788/YuruCamp2_02_-_TS_-_rock.ass.xz.zip)

TheOneric commented:
> That was on this file

Thanks, updated the previous table.

astiob commented:
At the very least it sounds like we’re never regressing, so this should be safe. The best-case gains are pretty impressive, too.

astiob commented:
Just one comment regarding secrecy: this is probably not what wyhash authors meant, but as you may know, Python and other languages that are used in server-side Web applications have to randomize their hash seeds and/or use special hashing algorithms to avoid DoS via carefully selected user-supplied hash table keys that is possible if the hashing algorithm is fully known (with all parameters) to an attacker. Now, of course, libass is not a server-side app, but theoretically someone might find it fun to create an ASS file that slows down rendering just through carefully chosen cache keys.

I have no idea what the “secret” name signifies in wyhash, though. Neither the code nor the paper makes it clear, either. I’m going to assume this is fine, but this choice of name baffles me.

This kind of lack of clarity/documentation (on wyhash upstream’s part, not rcombs’s) is my main/only lament about this whole ordeal.

rcombs commented:
To clarify, my ~45% figure comes from running the test file with an artificially-inflated frame rate (to increase the size of the actual test and reduce the proportion of overhead): `./profile/profile testfile.ass 10.57 10000 10.62`
master: 9.322s
this branch: 4.821s

rcombs commented:
Re-pushed with a few changes:
- Grabbed latest wyhash.h file verbatim (apart from trailing whitespace)
- Changed function name to be more generic
- Kept an init macro
- Switched to 64-bit integers (wyhash's native size)

astiob commented:
I welcome discussion on the merits and implementation of wyhash, but I’d rather that it happen elsewhere, i. e. upstream or in some other public forum where other interested parties may look.

TheOneric commented on libass/ass_utils.h:
```diff
@@ -180,26 +181,12 @@ static inline int double_to_d22(double x)
-    }
-
-    return hval;
+    static const uint64_t _wyp[5] = {0xa0761d6478bd642full, 0xe7037ed1a0b428dbull, 0x8ebc6af09c88c6e3ull, 0x589965cc75374cc3ull, 0x1d8e4e27c47d124full};
```
This exact value is already defined in the wyhash-header. Unless we decide to use a (possibly run-time generated) custom secret, this line can just be removed.

TheOneric commented:
Since this PR has last been updated, wyhash-upstream merged some changes to make the header compatible with more compilers, which I think warrants updating the header here. Also the `secret`-array now only needs 4 not 5 elements *(and the default `_wyp`-secret was temporarily removed but then restored)*.

I believe apart from some small and quick changes this PR was and is mostly ready for merge.
We just need to update the header and preferably also always define `WYHASH_LITTLE_ENDIAN` *(to use native endianess, since we don't care about reproducing the same hash on different machines)*.

rcombs commented:
Updated to latest upstream header as @TheOneric requested.

On the sign I pulled this test file from, this change reduces time spent in the hash function from 1930ms down to _68ms_, again for an overall improvement of around 40% in mpv playback, _including an HEVC 10-bit 4:4:4 decode_.
[magireco-sign.ass.zip](https://github.com/libass/libass/files/6972824/magireco-sign.ass.zip)

astiob commented:
Since the new hash is 64-bit, could someone benchmark this on a 32-bit architecture as well? (It doesn’t have to show a comparable order of improvement, but at least it shouldn’t become slower.)

TheOneric commented:
> could someone benchmark this on a 32-bit architecture as well?

Not the most precise tests, but it still clearly show there's no regression.
Test were done by rendering the MagiaRecord sample with profile from 2.42s 5.04s at 23.976fps; same script as in #469. Values are `ms` rounded to whole numbers (the values aren't even precise to 1ms anyway). All used gcc 8, but armv7 which used clang 12 instead.


|  Arch/ABI  |  CFLAGS                        |  master (1daecf4)  | wyhash (9662ef7)  | ≈wyhash/master |
|------------|--------------------------------|--------------------|-------------------|----------------|
|   amd64    | default                        |    755             |      260          |    34%         |
|   amd64    | -O3 -march=native -ffast-math  |    799             |      300          |    38%         |
|   i686     | -O2 -ftree-vectorize           |    768             |      484          |    63%         |
|armv7-softf*| default                        |   4405             |     2778          |    63%         |

* soft-float ABI but using a hardware FPU

Full run counts and timings attached: [wyhash_32bit_timings.zip](https://github.com/libass/libass/files/6979195/wyhash_32bit_timings.zip)

MrSmile commented on libass/ass_cache.c:
```diff
@@ -99,7 +99,7 @@ const CacheDesc bitmap_cache_desc = {
 
 // composite cache
-static uint32_t composite_hash(void *key, uint32_t hval)
+static uint64_t composite_hash(void *key, uint64_t hval)
```
I think it's better to use `typedef uint64_t ass_hash_t;` and use it. First commit looks like the perfect place to do it.

astiob replied:
@rcombs Any comment on the suggestion to introduce an `ass_hash_t`?

rcombs replied:
`_t` names are reserved for the system libraries as per POSIX and defining types with such names is undefined behavior, so I'm not going to add a new one of those. We could maybe do `ass_hashcode`?

astiob replied:
Right, of course; but do you like the idea? It seems pretty sensible to me.

astiob commented:
@TheOneric Sorry to bother you again, but could you perhaps rerun those 32-bit tests with `-DWYHASH_32BIT_MUM=1` and see if that makes it significantly faster still?

rcombs commented:
Note that in order to test with `WYHASH_32BIT_MUM`, you'll need to remove the `wyhashmap` function from the header; see https://github.com/wangyi-fudan/wyhash/issues/107.

TheOneric commented:
> rerun those 32-bit tests with `-DWYHASH_32BIT_MUM=1` and see if that makes it significantly faster still?

Same sample, same times and fps, same script but newer environment and slightly different CFLAGS. x86 tests now use gcc-10.2.1, armv7 uses clang-12.0.1. As before, this aren't the most precise tests which is especially true for the armv7 tests (which likely also suffers from some inaccuracies due to power and heat).

`master` means commit 88255b06dcf99ca37ae2493c1618dafb84837b2f, `wyhash` is this PR's commit 9662ef7d8e01ee69c6c04246e17e67e9fe60d9bf, but rebased on current master and `wyhash_32bitmum` is the same as before but with `WYHASH_32BIT_MUM` defined to `1`.


**amd64:**
**CFLAGS="`-O2 -ftree-vectorize -march=x86-64 -mtune=generic`"**

| **code** | **avg ms** | ratio to `master` |
|-------------|----------------|-----------------------|
| `master`                    |  743 | 100% |
| `wyhash`                   | 251 | 34% |
| `wyhash_32bitmum` | 276 | 37% |


**i686** (running on an amd64-CPU):
**CFLAGS="`-O2 -ftree-vectorize -march=i686 -mtune=i686`"**

| **code** | **avg ms** | ratio to `master` |
|-------------|----------------|-----------------------|
| `master`                    | 777 | 100% |
| `wyhash`                   | 436 | 56% |
| `wyhash_32bitmum` | 373 | 48% |


**armv7** (little-endian, EABI-softfloat backed by hardware FPU):
default CFLAGS

| **code** | **avg ms** | ratio to `master` |
|-------------|----------------|-----------------------|
| `master`                    | 3237 | 100% |
| `wyhash`                   | 2080 | 64% |
| `wyhash_32bitmum` | 2210 | 68% |

---

So it seems like `WYHASH_32BIT_MUM` is a bit faster on i686, but to be confident we'd need more precise and accurate tests. Taking the average runtimes at face value it would appear as if `WYHASH_32BIT_MUM` is *slower* on armv7 (a 32-bit architecture!) and amd64, but most likely the difference is within the precision of my testing so it doesn't say anything.
The only certain conlcusion I can draw from these tests is, that at least for the used CFLAGS and compilers the benefit of setting `WYHASH_32BIT_MUM`, if there even is any, is much smaller than the one obtained by switching to wyhash in the first place.

As before, here's the full run data, but bear in mind that the stddev of the runtimes only accounts for the purely statistical properties and not the surely existing systematic errors in the test procedure: [wyhash_32bit_mum_timings.zip](https://github.com/libass/libass/files/7165801/wyhash_32bit_mum_timings.zip)

MrSmile commented on libass/ass_cache.c:
```diff
@@ -99,7 +99,7 @@ const CacheDesc bitmap_cache_desc = {
 
 // composite cache
-static uint32_t composite_hash(void *key, uint32_t hval)
+static ass_hashcode composite_hash(void *key, ass_hashcode hval)
```
I think this is perfectly described by "give the hash a more generic name" and should go to the first commit.
In that case the last commit would be naturally localized in `ass_utils.h` as it should be.

TheOneric commented on libass/wyhash.h:
```diff
@@ -0,0 +1,266 @@
+    if(pos<size)	value[pos]++;	//	we process the vallue
+    else	cerr<<"the key does not exist\n";
+*/
+#ifdef	WYHASHMAP_WEAK_SMALL_FAST	// for small hashmaps whose size < 2^24 and acceptable collision
+typedef	uint32_t	wyhashmap_t;
+#else
+typedef	uint64_t	wyhashmap_t;
+#endif
```
Since you pointed out `_t` being reserved and the `WYHASH_32BIT_MUM` caveat and upstream already fixed that *(by removing the hashmap)*, could you update `wyhash.h` once more?

TheOneric commented:
I believe all nits and blockers should now be resolved. To recap:

 - While the `wyhash.h` header differs from our style, most were in favour of keeping it a verbatim copy to make future updates easier.
  - Compared to the first version from 8 months ago, `wyhash.h` now is no longer dependent on being compiled with gcc, clang or MSVC; no longer defines a `*_t` type; and fixed an error if `WYHASH_32BIT_MUM` is set.
 - We unconditionally define `WYHASH_LITTLE_ENDIAN` because we only care about reproducability on the same machine and this saves byte-swaps on big-endian machines
 - We checked i686 and armv7 to verify the new 64bit-hash isn't a performance regression *(in fact it is definitely a performance inprovement)*
 - Tests whether also defining `WYHASH_32BIT_MUM` on 32-bit arches was inconclusive, with the tendency being it's a tiny bit faster for i686 but a bit slower for armv7 — but it isn't sure if the difference might be within the variance of the test procedure. If it brings any improvement it is much smaller than that of switching to wyhash in the first place. So for now and until someone does more precise and accurate tests it's probably better to not define this.

 
For the future it was also suggested we could either generate a random hash-key instead of using the default or exposing a public API to set the hash key. However, there seemed to be issues with only some hash-key values being suitable for wyhash and it wasn't clear how to best deal with this *(which is why I don't think we need to already do this in this patch series)*.

It would be great if everyone could recheck this in case something is still missing and if not we can probably merge it.
